server/src/main/java/keywhiz/service/providers/ClientAuthenticator.java

```diff
@@ -200,23 +210,35 @@ public ClientAuthenticator(
         });
   }
 
+  /**
+   * Get a spiffe ID from a header, which must contain at most one entry and have
+   * that entry match the expected format for a SPIFFE URI. (If the header contains
+   * more than one entry, or the entry does not match SPIFFE URI format, Optional.empty()
+   * will be returned instead.)
+   *
+   * @param containerRequest the information included in a request to Keywhiz
+   * @param spiffeIdHeader the name of the header which should contain at most one SPIFFE URI
+   * @return the SPIFFE URI extracted from the header, or Optional.empty() if the header
+   * was missing or malformatted
+   */
   static Optional<URI> getSpiffeIdFromHeader(ContainerRequest containerRequest,
       String spiffeIdHeader) {
     List<String> spiffeIdHeaderValues =
         Optional.ofNullable(containerRequest.getRequestHeader(spiffeIdHeader)).orElse(List.of());
-    List<String> spiffeUriNames = spiffeUriNames(spiffeIdHeaderValues);
+    if (spiffeIdHeaderValues.size() > 1) {
+      logger.warn(
+          "Multiple header values are not allowed in the header {} that includes a SPIFFE URI (URIs: {})",
+          spiffeIdHeader, spiffeIdHeaderValues);
+      return Optional.empty();
+    }
 
+    List<String> spiffeUriNames = spiffeUriNames(spiffeIdHeaderValues);
     if (spiffeUriNames.isEmpty()) {
       logger.warn("No SPIFFE URI found from header {}", spiffeIdHeader);
       return Optional.empty();
     } else if (spiffeUriNames.size() > 1) {
       logger.warn("Got multiple SPIFFE URIs from header {}: {}", spiffeIdHeader, spiffeUriNames);
       return Optional.empty();
-    } else if (spiffeIdHeaderValues.size() > 1) {
-      logger.warn(
-          "Multiple URIs are not allowed in the header {} that includes a SPIFFE URI (URIs: {})",
-          spiffeIdHeader, spiffeIdHeaderValues);
-      return Optional.empty();
     }
```

Comment:
Yes, this isn't a behavior change; I found it easier to read if this check happened immediately after the values were extracted rather than some time after they'd been processed. There are tests in ClientAuthFactoryTest which pass multiple values in the headers (one for the XFCC header, one for the callerSpiffeId header).